### DISCUSSION TITLE
Change model via piping (objective function value)

### DISCUSSION THREAD

**Discussion by Percival24** [nlmixr2/nlmixr2#115] [Q&A] [answered]
_Hi! 

I am currently trying to change a model via piping. 
I am trying to add the covariate GENDER to a previous fit (fit_BASE). 
The parameter associated with GENDER is b1 which takes a value of zero as an initial value. 

What is surprising is the fact that the **final value** of the objective function of fit_BASE is not the same as the **initial value** of the objective function of my second run via piping (fit_GEN)... 
I don't really understand why I have something like a difference of 0.7 between the two objective functions. Besides, fit_GEN has the estimates of fit_BASE as initial estimates... 

Could anybody solve this mystery?_

```r
fit_GEN <- fit_BASE %>%
  model( { cl  <- exp(tcl + b1*GENDER +eta.cl) } ) %>%
  ini (b1 = 0) %>%
  nlmixr2(dataset,est="focei",
          foceiControl(outerOpt="nlminb",print=0),
          table=tableControl(cwres=T, npde=T))
```

_Take care,
Percival_

**Comment by mattfidler** [comment]
Hi @Percival24 

You may wish to fix the parameter by:

```r
fit_GEN <- fit_BASE %>%
  model( { cl  <- exp(tcl + b1*GENDER +eta.cl) } ) %>%
  ini (b1 = fix(0)) %>%
  nlmixr2(dataset,est="focei",
          foceiControl(outerOpt="nlminb",print=0),
          table=tableControl(cwres=T, npde=T))
```

for this test.  

However, even if they have the *same* population population parameters, if they have different between subject variabilites they will have different objective functions.  So, check to see if the problem has different `eta` values to see why the objective function is different.

**Reply by Percival24** [reply]
Thanks. But I checked that. Let me show more clearly what I have on my screen.

**Comment by Percival24** [comment]
Then end of fitbase is the exact beginning of fit1.
But the objective functions have different values... 31445.55  and 31448.66 

```r
fitbase <- nlmixr2(model.1cpt.ode,pool123,seed = 99,est="focei",foceiControl(outerOpt="nlminb",print=0),table=tableControl(cwres=T, npde=T))

last = length(fitbase$parHist[,1])
fitbase$parHist[last,]
```

```r
### R_Console
 fitbase$parHist[last,]
 iter      objf      tclnm      tclm      tv      tvm      tv2       tq      tqm    lagtime  prop.err propm.err       o1        o2       o3       o4       o5       o6
 101   34  31445.55  3.43485 0.6104284 4.83747 5.593781 4.295547 3.807615 3.064538 0.09977438 0.4205748 0.3414423 2.054087 -1.249223 1.296046 2.775941 2.786186 2.719084_

fit1 <- fitbase %>% 
  model( { clnm  <- exp(tclnm + b1*Dobu + eta.clnm) } ) %>%
  ini (b1 = 0) %>%
  nlmixr2(pool123,est="focei",
          foceiControl(outerOpt="nlminb",print=0),
          table=tableControl(cwres=T, npde=T))

fit1$parHist[1,]
```

**_R_Console
```r
 fit1$parHist[1,]
 iter     objf      tclnm      tclm      tv      tvm      tv2       tq      tqm    lagtime  prop.err propm.err b1       o1        o2       o3       o4       o5       o6
 2    1   31448.66 3.43485 0.6104284 4.83747 5.593781 4.295547 3.807615 3.064538 0.09977438 0.4205748 0.3414423  0 2.054087 -1.249223 1.296046 2.775941 2.786186 2.719084_**
 
```

**Reply by mattfidler** [reply]
The default is to reset all the `eta` values to zero for model piping.  This means it is likely these have different outcomes at the beginning.  The model is not resumed with the prior `eta` values, though that could be added as a feature in the future.

This matches what has been done in modeling tools like NONMEM and Monolix, they typically do not pre-specify the `eta` values (though I believe you can in NONMEM, I am unsure if you can in Monolix).

So, the objective functions are not terribly different, which is in line with my hypothesis.  There is another set of values picked out for the best objective function on the first step than with the last step of the prior optimization.

**Reply by mattfidler** [reply]
Your reply didn't show that the ETA parameters were the same or different, so it is unclear from your post if this is correct.

I believe you can access the ETA values with `$eta`.  You can also skip the step and try  `est="posthoc"` so you can more clearly see the difference.  In the `"posthoc"` estimation step the only thing that is optimized is the `eta` values themselves.

**Comment by Percival24** [comment]
I understand.  
So far, it is impossible to use the piping with the previous etas estimates. 
It's unfortunate because to know if a covariate has to be added to the model or not we have to compare two objective functions. 
If the first objective function decreases by (at least) 3.84, then we add the covariate. 
3.84 is relatively small compare to objective function value like 31445.55.... That why we need to be accurate during the piping. 

I researched the etas in fitbase and fit1.  
I could only find the final estimates of these parameters and not all the optimization steps like shown before. 
Bellow are the final etas for the two fits.

**Comment by Percival24** [comment]
```r fitbase$eta[,1:2]
ID        eta.clnm
1  E1-01  0.194336473465
2  E1-02  0.404162109778
3  E1-03  0.349092131237
4  E1-04 -0.007794828480
5  E1-05  0.220784436023
6  E1-06  0.066562649007
7  E1-07  0.336928280109
8  E1-08  0.197182207436
9  E1-09  0.557106427481
10 E1-10  0.480403582595
11 E1-11  0.373261620833
12 E1-12  0.473901533456
13 E1-13  0.321228947888
14 E1-14  0.264748009268
15 E1-15  0.527625553048
16 E1-16  0.390406031198
17 E2-01 -0.028969758758
18 E2-02 -0.827663799590
19 E2-03 -0.792789020492
20 E2-04  0.007517809203
21 E2-05  0.391755039725
22 E2-06 -0.163576964364
23 E2-07 -0.440310416809
24 E2-08 -0.590655625540
25 E2-09 -0.370326675178
26 E2-10 -0.413860094289
27 E2-11 -0.199975330809
28 E2-12 -0.234151097420
29 E3-01  0.538394122139
30 E3-02  0.103479606005
31 E3-03  0.176145050709
32 E3-04  0.007575506825
33 E3-05  0.206319832922
34 E3-06  0.331621560454
35 E3-08  0.273472028359
36 E3-09  0.233549239430
37 E3-10  0.075407716382
38 E3-12  0.327508516652
39 E3-13  0.067521494927
40 E3-14  0.083117376436
```

**Comment by Percival24** [comment, marked as answer]
```r
fit1$eta[,1:2]
ID         eta.clnm
1  E1-01  0.1972839888179
2  E1-02  0.4052014011313
3  E1-03  0.3441006022338
4  E1-04 -0.0205932979943
5  E1-05  0.1984915097270
6  E1-06  0.0605827459473
7  E1-07  0.3069786106801
8  E1-08  0.1916509442910
9  E1-09  0.5508381549362
10 E1-10  0.4744701174127
11 E1-11  0.3664667299211
12 E1-12  0.4683123246814
13 E1-13  0.3136001447208
14 E1-14  0.2530760771880
15 E1-15  0.5227565941289
16 E1-16  0.3797068801798
17 E2-01 -0.0848199923500
18 E2-02 -0.8306272282715
19 E2-03 -0.7956844892765
20 E2-04 -0.0077417805859
21 E2-05  0.3627357859914
22 E2-06 -0.1258762364686
23 E2-07 -0.4438873857666
24 E2-08 -0.5320327784873
25 E2-09 -0.3480346748383
26 E2-10 -0.4295661177151
27 E2-11 -0.2157107739208
28 E2-12 -0.2414214099228
29 E3-01  0.5359181346544
30 E3-02  0.1016649453386
31 E3-03  0.1726291744747
32 E3-04  0.0002481307201
33 E3-05  0.2004843988764
34 E3-06  0.3281820901047
35 E3-08  0.2714783278413
36 E3-09  0.2281272301118
37 E3-10  0.0680875187148
38 E3-12  0.3235728014038
39 E3-13  0.0655607750910
40 E3-14  0.0790531393424
```

**Reply by mattfidler** [reply]
It is actually standard practice to *not* update the etas on checking for covariates, but let them all reset to zero.

You actually *can* supply the `eta` values to nlmixr2 focei routines by supplying `etaMat` to the `foceiControl()` control item.  But again, this is not standard practice when looking for significant covariates.

**Reply by mattfidler** [reply]
And the `eta` values appear to be different (and unsurprisingly the objective functions are then different)

**Reply by mattfidler** [reply]
You cannot currently supply `eta` values to `saem` though.

**Reply by Percival24** [reply]
Thanks a lot, @mattfidler !!!! 
It is really helpful :)

**Reply by mattfidler** [reply]
No problem.
